### DISCUSSION TITLE
How to register a (repeatedly) sampled random tensor?

### DISCUSSION THREAD

**Discussion by RylanSchaeffer** [Lightning-AI/pytorch-lightning#14131] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
In my code, with each batch, I sample a random tensor. However, when I try porting to GPU, I get an error that I'm trying to multiply a tensor on CPU with a tensor on GPU. The [documentation](https://pytorch-lightning.readthedocs.io/en/latest/accelerators/accelerator_prepare.html) states

> The LightningModule knows what device it is on. You can access the reference via self.device. Sometimes it is necessary to store tensors as module attributes. However, if they are not parameters they will remain on the CPU even if the module gets moved to a new device. To prevent that and remain device agnostic, register the tensor as a buffer in your modules’ __init__ method with register_buffer().

The example given is:

```
class LitModel(LightningModule):
    def __init__(self):
        ...
        self.register_buffer("sigma", torch.eye(3))
        # you can now access self.sigma anywhere in your module
```

Using this example, suppose I want to randomly sample a 3x3 matrix `sigma` with each batch. How do I properly register this tensor?

**Comment by RylanSchaeffer** [comment, marked as answer]
I just found this answer from 2020: https://stackoverflow.com/questions/63660624/normal-distribution-sampling-in-pytorch-lightning

Is the best solution currently to specify the device?

**Reply by akihironitta** [reply]
> In my code, with each batch, I sample a random tensor. However, when I try porting to GPU, I get an error that I'm trying to multiply a tensor on CPU with a tensor on GPU.

As the thread suggests, you need to move your tensor to GPU manually with `self.device` in each iteration.

**Comment by RylanSchaeffer** [comment]
That's the 2020 solution. Is there not a more modern approach?

**Reply by akihironitta** [reply]
There's no "modern" approach I could think of, and I'd rather say that PyTorch and PyTorch Lightning are stable enough to keep the same APIs for a few years now :)

Or is there any particular feature you'd like to see in PL?

**Comment by RylanSchaeffer** [comment]
Ok cool thank you!

Cheers,
Rylan Schaeffer
